File: backend_samples/library/python/AzureAIVisionFaceDeviceAttestation/azure_ai_vision_face_deviceattestation/ios/verify.py
"""
Top-level iOS App Attest verifier — sequences the attestation phases.
"""

from __future__ import annotations

import logging
import time
from dataclasses import asdict
from datetime import datetime, timezone

from cryptography import x509
from cryptography.hazmat.primitives import hashes, serialization

from ..config import get_attestation_config
from ..logging import track_event, track_exception
from ..cert_utils import pem_to_der
from .attestation_phases import (
    parse_attest_auth_data,
    parse_ios_attest_envelope,
    validate_ios_chain_validity_dates,
    verify_aaguid_policy,
    verify_app_id_rp_id_hash,
    verify_assertion_against_auth_cert,
    verify_attestation_nonce_binding,
    verify_credential_id_binds_pub_key,
    verify_ios_x5c_chain,
)
from .types import (
    AppAttestVerdict,
    AssertionSummary,
    CredCertSummary,
    IntermediateCertSummary,
    IosPhaseFail,
)

log = logging.getLogger("ios.verify")


def _cert_to_dict_summary(cert: x509.Certificate) -> tuple[str, str, str, str, str]:
    return (
        cert.subject.rfc4514_string(),
        cert.issuer.rfc4514_string(),
        format(cert.serial_number, "X"),
        cert.not_valid_before_utc.isoformat(),
        cert.not_valid_after_utc.isoformat(),
    )


async def verify_ios_auth(message_data: dict, attest_json: str) -> dict:
    start = time.time() * 1000
    base_props = {
        "platform": "ios",
        "clientId": message_data.get("clientId"),
        "challengeHashPrefix": (message_data.get("challengeHash") or "")[:8],
        "attestJsonLength": len(attest_json),
    }
    track_event("IosAuth.VerifyStart", base_props)

    def fail_verify(reason: str, message: str, extras=None, extra_props=None) -> dict:
        duration = time.time() * 1000 - start
        log.error("[iOSAuth] VerifyFail %s %s", reason, message)
        track_event(
            "IosAuth.VerifyFail",
            {**base_props, "reason": reason, "message": message, **(extra_props or {})},
            {"durationMs": duration},
        )
        result = {
            "verified": False,
            "platform": "ios",
            "message": message,
            "timestamp": datetime.now(timezone.utc).isoformat(),
        }
        if extras:
            result.update(extras)
        return result

    try:
        debug_mode = get_attestation_config().debug_mode
        try:
            env = parse_ios_attest_envelope(attest_json, base_props)
            parsed_token = env["parsedToken"]
            parsed_assertion = env["parsedAssertion"]

            validate_ios_chain_validity_dates(
                parsed_token.credCertDer, parsed_token.intermediateDer, base_props
            )
            verify_ios_x5c_chain(parsed_token.credCertDer, parsed_token.intermediateDer)

            warnings: list[str] = []

            ad = parse_attest_auth_data(parsed_token.authData)
            rp_id_hash = ad["rpIdHash"]
            sign_count = ad["signCount"]
            aaguid = ad["aaguid"]
            credential_id = ad["credentialId"]

            app_id_res = verify_app_id_rp_id_hash(rp_id_hash)
            expected_app_id = app_id_res["expectedAppId"]

            verify_aaguid_policy(aaguid, debug_mode, warnings)

            cred_bind = verify_credential_id_binds_pub_key(parsed_token.credCertDer, credential_id, base_props)
            cred_cert_pub_key = cred_bind["credCertPubKey"]

            nonce_res = verify_attestation_nonce_binding(parsed_token.credCertDer, parsed_token.authData, message_data)
            cert_nonce = nonce_res["certNonce"]
            challenge_bytes = nonce_res["challengeBytes"]

            assertion_res = verify_assertion_against_auth_cert(
                parsed_assertion=parsed_assertion,
                cred_cert_pub_key=cred_cert_pub_key,
                attest_rp_id_hash=rp_id_hash,
                attest_sign_count=sign_count,
                auth_cert_pem=message_data["publicCert"],
                base_props=base_props,
            )

            duration = time.time() * 1000 - start
            cred_cert = x509.load_der_x509_certificate(parsed_token.credCertDer)
            int_cert = x509.load_der_x509_certificate(parsed_token.intermediateDer)
            import hashlib as _h
            cred_cert_thumbprint = _h.sha256(parsed_token.credCertDer).hexdigest()
            aaguid_utf8 = aaguid.rstrip(b"\x00").decode("utf-8", errors="replace")
            cred_cert_pem = cred_cert.public_bytes(serialization.Encoding.PEM).decode("ascii")

            cred_sub, cred_iss, cred_sn, cred_vf, cred_vt = _cert_to_dict_summary(cred_cert)
            int_sub, int_iss, int_sn, int_vf, int_vt = _cert_to_dict_summary(int_cert)

            verdict = AppAttestVerdict(
                fmt=parsed_token.fmt,
                rpIdHash=rp_id_hash.hex(),
                appId=expected_app_id,
                aaguid=aaguid_utf8,
                flags=ad["flags"],
                signCount=sign_count,
                credentialId=credential_id.hex(),
                credentialIdMatchesPubKey=True,
                credCert=CredCertSummary(cred_sub, cred_iss, cred_sn, cred_vf, cred_vt, cred_cert_thumbprint),
                credCertPem=cred_cert_pem,
                intermediateCert=IntermediateCertSummary(int_sub, int_iss, int_sn, int_vf, int_vt),
                receiptLength=parsed_token.receiptLength,
                authDataLength=len(parsed_token.authData),
                nonceExtension=cert_nonce.hex(),
                expectedClientDataHash=challenge_bytes.hex(),
                authCertThumbprint=assertion_res["authCertThumbprintHex"],
                nonceVerified=True,
                assertion=AssertionSummary(
                    authenticatorDataLength=len(assertion_res["assertionAuthData"]),
                    rpIdHash=assertion_res["assertionRpIdHash"].hex(),
                    flags=assertion_res["assertionFlags"],
                    signCount=assertion_res["assertionSignCount"],
                    signatureLength=len(parsed_assertion.signature),
                    signatureEncoding=assertion_res["signatureEncodingUsed"],
                    expectedClientDataHash=assertion_res["authCertThumbprintHex"],
                    signatureVerified=True,
                ),
                challengeBinding=(
                    "attest: sha256(authData || challengeHashBytes);  "
                    "assert: ECDSA-SHA256(credCertPubKey, nonce = sha256(authenticatorData || sha256(authCertDER)))"
                ),
            )

            track_event(
                "IosAuth.VerifySuccess",
                {
                    **base_props,
                    "rpIdHashPrefix": rp_id_hash.hex()[:12],
                    "credCertSubject": cred_cert.subject.rfc4514_string(),
                    "aaguid": aaguid_utf8,
                    "flags": ad["flags"],
                    "attestSignCount": sign_count,
                    "assertSignCount": assertion_res["assertionSignCount"],
                    "warningCount": len(warnings),
                },
                {"durationMs": duration},
            )

            result = {
                "verified": True,
                "platform": "ios",
                "message": "iOS App Attest verified successfully",
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "challengeHash": message_data.get("challengeHash"),
                "clientId": message_data.get("clientId"),
                "attestJsonLength": len(attest_json),
                "chainLength": 2,
                "rootCA": "Apple App Attestation Root CA",
                "appAttestVerdict": asdict(verdict),
            }
            if warnings:
                result["warnings"] = warnings
            return result
        except IosPhaseFail as f:
            return fail_verify(f.reason, f.message, f.extras, f.extra_props)
    except Exception as err:
        log.exception("[iOSAuth] Unexpected error")
        track_exception(err, {"source": "verify_ios_auth.unexpected", **base_props})
        return fail_verify("UNEXPECTED_ERROR", f"iOS attestation verification error: {err}")

File: backend_samples/library/javascript/AzureAIVisionFaceDeviceAttestation/src/ios/verify.ts
/**
 * Top-level iOS App Attest orchestrator.
 *
 * Each verification step is one named phase in `attestation_phases.ts`; this
 * file is just the sequencer + telemetry adapter. Phases return their result
 * payload on success, or an `IosPhaseFail` carrying the reason code + extras
 * — `failVerify` turns that into the AuthVerificationResult the caller
 * (auth_verification.verifyAuthBySystem) already expects.
 *
 *   attestation:
 *     {
 *       "fmt":     "apple-appattest",
 *       "attStmt": { "x5c": [credCert, intermediateCert], "receipt": <bytes> },
 *       "authData": <bytes>,
 *     }
 *
 *   assertion:
 *     {
 *       "signature":         <ECDSA-P256-SHA256 DER bytes>,
 *       "authenticatorData": <37+ bytes: rpIdHash || flags || signCount || ...>
 *     }
 *
 *   attestation clientDataHash  = Buffer.from(messageData.challengeHash, 'hex')
 *   expectedAttestNonce         = SHA-256( authData || attestation clientDataHash )
 *   expectedAttestNonce        == nonce OCTET STRING in credCert ext 1.2.840.113635.100.8.2
 *
 *   assertion clientDataHash    = Buffer.from(sha256(authPublicCertDER).hex, 'hex')
 *   nonce                       = SHA-256( assertion.authenticatorData || assertion clientDataHash )
 *   ECDSA_with_SHA256_verify( credCertPubKey, signature, nonce )  ==  true
 *
 * Why two proofs instead of one folded hash:
 *   - Attestation proves Apple-hardware provenance against the server challenge.
 *   - Assertion proves the auth cert that's being registered belongs to the
 *     same attested key. Splitting the two lets us audit and reason about each
 *     independently, and matches Apple's intended use of the two APIs.
 */
import crypto from 'crypto';
import type { AuthVerificationResult, AttestationMessageData } from '../auth_verification';
import { trackEvent, trackException } from '../logging';
import { getAttestationConfig } from '../config';
import { IosPhaseFail, type IosBaseProps, type AppAttestVerdict } from './types';
import {
  parseIosAttestEnvelope,
  verifyIosX5cChain,
  validateIosChainValidityDates,
  parseAttestAuthData,
  verifyAppIdRpIdHash,
  verifyAaguidPolicy,
  verifyCredentialIdBindsPubKey,
  verifyAttestationNonceBinding,
  verifyAssertionAgainstAuthCert,
} from './attestation_phases';

/**
 * Verify Apple App Attest attestation + assertion.
 *
 * @param messageData - Parsed message object (challengeHash, clientId, system, publicCert)
 * @param attestJson  - JSON string `{ "attestation": "<base64 CBOR>", "assertion": "<base64 CBOR>" }`
 * @returns Verification result.
 */
export async function verifyiOSAuth(
  messageData: AttestationMessageData,
  attestJson: string,
): Promise<AuthVerificationResult> {
  const startTime = Date.now();
  const baseProps: IosBaseProps = {
    platform: 'ios',
    clientId: messageData.clientId,
    challengeHashPrefix: messageData.challengeHash?.slice(0, 8),
    attestJsonLength: attestJson.length,
  };

  trackEvent('IosAuth.VerifyStart', baseProps);

  const failVerify = (
    reason: string,
    message: string,
    extras: Partial<AuthVerificationResult> = {},
    extraProps: Record<string, unknown> = {},
  ): AuthVerificationResult => {
    const durationMs = Date.now() - startTime;
    trackEvent(
      'IosAuth.VerifyFail',
      { ...baseProps, reason, message, ...extraProps },
      { durationMs },
    );
    return {
      verified: false,
      platform: 'ios',
      message,
      timestamp: new Date().toISOString(),
      ...extras,
    };
  };

  const fail = (f: IosPhaseFail) => failVerify(f.reason, f.message, f.extras, f.extraProps);

  try {
    const debugMode = getAttestationConfig().debugMode;

    // 1. parse JSON + CBOR envelopes
    const env = parseIosAttestEnvelope(attestJson, baseProps);
    if (env instanceof IosPhaseFail) return fail(env);
    const { parsedToken, parsedAssertion } = env;
    const { fmt, authData, credCertDer, intermediateDer, receiptLength, receipt } = parsedToken;

    const validityFail = validateIosChainValidityDates(credCertDer, intermediateDer, baseProps);
    if (validityFail) return fail(validityFail);
    const chainFail = await verifyIosX5cChain(credCertDer, intermediateDer);
    if (chainFail) return fail(chainFail);
    const warnings: string[] = [];

    // 4. authData layout
    const authDataRes = parseAttestAuthData(authData);
    if (authDataRes instanceof IosPhaseFail) return fail(authDataRes);
    const { rpIdHash, flags, signCount, aaguid, credIdLen, credentialId } = authDataRes;

    // 5. app id (rpIdHash binding)
    const appIdRes = verifyAppIdRpIdHash(rpIdHash);
    if (appIdRes instanceof IosPhaseFail) return fail(appIdRes);
    const { expectedAppId } = appIdRes;

    // 6. aaguid policy (prod/dev gating)
    const aaguidFail = verifyAaguidPolicy(aaguid, debugMode, warnings);
    if (aaguidFail) return fail(aaguidFail);

    trackEvent('IosAuth.AuthDataParsed', {
      rpIdHashPrefix: rpIdHash.toString('hex').slice(0, 16),
      flags,
      signCount,
      aaguid: aaguid.toString('utf8').replace(/\0+$/, ''),
      credentialIdLength: credIdLen,
      credentialIdPrefix: credentialId.toString('hex').slice(0, 16),
      expectedAppId,
      appIdMatched: true,
    });

    // 7. credentialId == SHA-256(uncompressed EC point)
    const credBindRes = verifyCredentialIdBindsPubKey(credCertDer, credentialId, baseProps);
    if (credBindRes instanceof IosPhaseFail) return fail(credBindRes);
    const { credCertPubKey } = credBindRes;

    // 8. attestation nonce binding to session challenge
    const nonceRes = verifyAttestationNonceBinding(credCertDer, authData, messageData);
    if (nonceRes instanceof IosPhaseFail) return fail(nonceRes);
    const { certNonce, challengeBytes } = nonceRes;

    // 9. assertion proves the auth cert belongs to the attested key
    const assertionRes = verifyAssertionAgainstAuthCert({
      parsedAssertion,
      credCertPubKey,
      attestRpIdHash: rpIdHash,
      attestSignCount: signCount,
      // messageData.publicCert is guaranteed by verifyAttestationNonceBinding
      authCertPem: messageData.publicCert!,
      baseProps,
    });
    if (assertionRes instanceof IosPhaseFail) return fail(assertionRes);
    const {
      assertionAuthData,
      assertionRpIdHash,
      assertionFlags,
      assertionSignCount,
      signatureEncodingUsed,
      authCertThumbprintHex,
    } = assertionRes;

    trackEvent('IosAuth.AppAttestVerified', {
      ...baseProps,
      assertionSignCount,
      signatureEncoding: signatureEncodingUsed,
    });

    // ---- Build the decoded verdict for persistence. ----
    const durationMs = Date.now() - startTime;
    const credCert = new crypto.X509Certificate(credCertDer);
    const intCert = new crypto.X509Certificate(intermediateDer);
    const credCertThumbprint = crypto.createHash('sha256').update(credCertDer).digest('hex');
    const aaguidUtf8 = aaguid.toString('utf8').replace(/\0+$/, '');

    const verdict: AppAttestVerdict = {
      fmt,
      rpIdHash: rpIdHash.toString('hex'),
      appId: expectedAppId,
      aaguid: aaguidUtf8,
      flags,
      signCount,
      credentialId: credentialId.toString('hex'),
      credentialIdMatchesPubKey: true,
      credCert: {
        subject: credCert.subject,
        issuer: credCert.issuer,
        serialNumber: credCert.serialNumber,
        validFrom: credCert.validFrom,
        validTo: credCert.validTo,
        thumbprint: credCertThumbprint,
      },
      credCertPem: credCert.toString(),
      intermediateCert: {
        subject: intCert.subject,
        issuer: intCert.issuer,
        serialNumber: intCert.serialNumber,
        validFrom: intCert.validFrom,
        validTo: intCert.validTo,
      },
      receiptLength,
      receipt: receipt ? receipt.toString('base64') : null,
      authDataLength: authData.length,
      nonceExtension: certNonce.toString('hex'),
      expectedClientDataHash: challengeBytes.toString('hex'),
      authCertThumbprint: authCertThumbprintHex,
      nonceVerified: true,
      assertion: {
        authenticatorDataLength: assertionAuthData.length,
        rpIdHash: assertionRpIdHash.toString('hex'),
        flags: assertionFlags,
        signCount: assertionSignCount,
        signatureLength: parsedAssertion.signature.length,
        signatureEncoding: signatureEncodingUsed,
        expectedClientDataHash: authCertThumbprintHex,
        signatureVerified: true,
      },
      challengeBinding:
        'attest: sha256(authData || challengeHashBytes);  assert: ECDSA-SHA256(credCertPubKey, nonce = sha256(authenticatorData || sha256(authCertDER)))',
    };

    trackEvent(
      'IosAuth.VerifySuccess',
      {
        ...baseProps,
        rpIdHashPrefix: rpIdHash.toString('hex').slice(0, 12),
        credCertSubject: credCert.subject,
        aaguid: aaguidUtf8,
        flags,
        attestSignCount: signCount,
        assertSignCount: assertionSignCount,
        warningCount: warnings.length,
      },
      { durationMs },
    );

    return {
      verified: true,
      platform: 'ios',
      message: 'iOS App Attest verified successfully',
      timestamp: new Date().toISOString(),
      challengeHash: messageData.challengeHash,
      clientId: messageData.clientId,
      attestJsonLength: attestJson.length,
      chainLength: 2,
      rootCA: 'Apple App Attestation Root CA',
      appAttestVerdict: verdict,
      ...(warnings.length > 0 && { warnings }),
    };
  } catch (error) {
    trackException(error, { source: 'verifyiOSAuth.unexpected', ...baseProps });
    return failVerify(
      'UNEXPECTED_ERROR',
      `iOS attestation verification error: ${error instanceof Error ? error.message : 'unknown error'}`,
    );
  }
}
